cborg/src/Codec/CBOR/Decoding.hs

```diff
@@ -986,7 +987,7 @@ type ByteOffset = Int64
 -- @since 0.2.2.0
 peekByteOffset :: Decoder s ByteOffset
 peekByteOffset = Decoder (\k -> return (PeekByteOffset (\off# -> k (I64#
-#if MIN_VERSION_base(4,17,0)
+#if MIN_VERSION_base(4,17,0) && !defined(ARCH_32bit)
```

Comment:
Surely this can't be quite the right logic, because 32bit arches prior to some ghc version did need `intToInt64#` didn't they?

Reply:
`PeekByteOffset` always provides an `Int64#` on 32bit:
https://github.com/well-typed/cborg/blob/c8013b3474d876f4da56c869d57e3f3ac7f42dc6/cborg/src/Codec/CBOR/Decoding.hs#L191-L195 so this seems correct to me.

Reply:
> PeekByteOffset always provides an Int64# on 32bit

which would mean this logic here is wrong. The logic currently says "use intToInt64 if base >= 4.17 and only on 64bit arches".

But what you're saying above is that it should be "always use intToInt64 on 32bit arches", so the combo would be

```
#if MIN_VERSION_base(4,17,0) || defined(ARCH_32bit)
```
so that we use intToInt64 on 32bit arches (for all base versions), and we also use it on all arches from base 4.17 onwards.

Did we check this patch actually compiles on 32bit arches prior to ghc 9.2?

Reply:
> > PeekByteOffset always provides an Int64# on 32bit
> 
> which would mean this logic here is wrong. The logic currently says "use intToInt64 if base >= 4.17 and only on 64bit arches".
> 
> But what you're saying above is that it should be "always use intToInt64 on 32bit arches", so the combo would be
> 
> ```
> #if MIN_VERSION_base(4,17,0) || defined(ARCH_32bit)
> ```
> 
> so that we use intToInt64 on 32bit arches (for all base versions), and we also use it on all arches from base 4.17 onwards.

I think the existing logic is exactly right (it also compiles on all GHC versions >=8.4), as seen eg by a case distinction.

The purpose of this code is to fill the hole `_f` in `PeekByteOffset $ \off# -> k (I64# (_f off#))` (where `k` has input type `ByteOffset ~ Int64`).

 - On 32bit:
   We have `off# :: Int64#` and `I64# :: Int64# -> Int64` on all GHCs. Hence, we can fill `_f = id`.
 - On 64bit:
   We have `off# :: Int#`.
    - On GHC < 9.4, we have `I64# :: Int# -> Int64`.
      Hence, we can fill `_f = id`.
    - Since GHC >= 9.4, we have `I64# :: Int64# -> Int64` (FTR: [corresponding section in the migration guide](https://gitlab.haskell.org/ghc/ghc/-/wikis/migration/9.4#word64int64-use-word64int64) for 9.4).
      Hence, we can fill `_f = intToInt64# :: Int# -> Int64#`

Note that we can fill `_f = id` *except* when we are on GHC >=9.4 and on 64bit, which is exactly the existing condition.

Would be neat if there were some more principled/locally understandable alternative to the current CPP maze :sweat_smile: 

> Did we check this patch actually compiles on 32bit arches prior to ghc 9.2?

Yes, to double check once again, I pushed to a branch in my fork with the second commit ([Support GHC >=9.2 on 32bit](https://github.com/well-typed/cborg/pull/322/commits/299bb3e6918f7e835c6ecac32efe4fada7514d8f)) removed: https://github.com/amesgen/cborg/actions/runs/6183531233
One can see that i686 compilation with all GHCs from 8.4 to 9.0 succeeds (as expected, compilation on >=9.2 fails as the respective commit was removed).
Also, here is the CI run that everything works on all GHCs with all three commits: https://github.com/amesgen/cborg/actions/runs/6138762687